### PR TITLE
Tortoise

### PR DESCRIPTION
**Pull request by antonlerner**

### CODE REVIEW

File: core/block.go [hunk 1]
```diff
@@ -0,0 +1,53 @@
+package core
```
Comment (zalmen) [inline core/block.go]: rename to "mesh", "core" is too general

File: core/block.go [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+import (
+	"time"
+	"github.com/google/uuid"
+)
+
+type BLockId uint32//uuid.UUID
```
Comment (zalmen) [inline core/block.go]: remove comment

File: core/block.go [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+}
+
+func (l *Layer) AddBlock(block *Block){
+	block.layerNum = l.layerNum
+	l.blocks = append(l.blocks, *block)
+}
+
+func NewLayer() *Layer{
+	l := Layer{
+		blocks: make([]Block,0),
```
Comment (zalmen) [inline core/block.go]: why is it set to zero? this will cause every append of a block to copy the entire slice

File: core/block.go [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+		layerNum: layerCounter,
+	}
+	layerCounter++
```
Comment (zalmen) [inline core/block.go]: Not sure what is the "thread-safty" nature of this code. Should it be protected by a lock?

File: core/block.go (hunk 2 above)
Comment (zalmen) [reply core/block.go]: and fix capitalisation

File: core/tortoise.go [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+
+func (alg *Algorithm) LayerVotingAvg() uint64 {
+	return 30
+}
+
+func (alg *Algorithm) IsTortoiseValid(originBlock *Block, targetBlock BLockId, targetBlockIdx uint64, visibleBlocks bitarray.BitArray) bool {
+	//log.Info("calculating votes for: %v on target id: %v",alg.block2Id[originBlock.id], targetBlockId)
+	voteFor, voteAgainst := alg.countTotalVotesForBlock(targetBlockIdx, visibleBlocks)
+
+	/*if err != nil {
```
Comment (zalmen) [inline core/tortoise.go]: remove

File: core/tortoise.go [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+	}
+	return &blockMap, &visibilityMap
+}
+
+func (alg *Algorithm) countTotalVotesForBlock(targetIdx uint64, visibleBlocks bitarray.BitArray) (uint64, uint64){
+	//targetId := alg.block2Id[target.id]
+	var posVotes, conVotes uint64 = 0,0
+	targetLayer := alg.visibilityMap[targetIdx].layer
+	ln := len(alg.block2Id)
+	for blockIdx:=0; blockIdx< ln; blockIdx++{//_, blockIdx := range alg.block2Id { possible bug what if there is an id > len(alg.block2id)
```
Comment (zalmen) [inline core/tortoise.go]: remove comment

Comment (zalmen) [reply core/tortoise.go]: There are few places in the code where you left commented-out code. Please remove them

**Review by zalmen — APPROVED** [review]

File: core/block.go (hunk 3 above)
Comment (antonlerner) [reply core/block.go]: This will be changed later anyways. I'd like to keep this since it allows me to push items one by one easily

**Pull request merged**